File: dump_tmpl_ast.dart
import 'dart:convert';
import 'dart:io';
import 'package:vue_sfc_parser/template/dom_options.dart';
import 'package:vue_sfc_parser/template/tmpl_ast.dart';
import 'package:vue_sfc_parser/template/tmpl_parser.dart';

String enc(String t) => jsonEncode(t);
String loc(TmplNode n) => '@${n.loc.start.offset}-${n.loc.end.offset}';

String dumpProp(TmplNode p) {
  if (p is AttributeNode) {
    final v = p.value;
    return 'ATTR ${p.name} ${v != null ? enc(v.content) : "noval"} ${loc(p)}';
  }
  final d = p as DirectiveNode;
  final a = d.arg;
  final arg = a is SimpleExpression
      ? '${a.content}${a.static_ ? ":s" : ":d"}'
      : '-';
  final e = d.exp;
  final exp = e is SimpleExpression ? e.content : '-';
  final mods = d.modifiers.map((m) => m.content).join(',');
  return 'DIR ${d.name} raw=${d.rawName} arg=$arg exp=${enc(exp)} mods=[$mods] ${loc(p)}';
}

void walk(TmplNode n, int d, List<String> lines) {
  final pad = '  ' * d;
  if (n is ElementNode) {
    lines.add(
        '${pad}EL ${n.tag} ns=${n.ns} tagType=${n.tagType} self=${n.isSelfClosing ? 1 : 0} ${loc(n)}');
    for (final p in n.props) {
      lines.add('$pad  ${dumpProp(p)}');
    }
    for (final c in n.children) {
      walk(c, d + 1, lines);
    }
  } else if (n is TextNode) {
    lines.add('${pad}TEXT ${enc(n.content)} ${loc(n)}');
  } else if (n is CommentNode) {
    lines.add('${pad}COMMENT ${enc(n.content)} ${loc(n)}');
  } else if (n is InterpolationNode) {
    final c = n.content;
    lines.add(
        '${pad}INTERP ${enc(c is SimpleExpression ? c.content : "<compound>")} ${loc(n)} exp@${c.loc.start.offset}-${c.loc.end.offset}');
  }
}

void main(List<String> args) {
  final list = jsonDecode(File('samples_tmpl.json').readAsStringSync()) as List;
  final e = list.firstWhere((x) => x['name'] == args[0]);
  final sfc = e['sfc'] as String;
  final m = RegExp(r'<template>([\s\S]*)</template>').firstMatch(sfc)!;
  final source = m.group(1)!;
  final root = baseParse(source, domParserOptions());
  final lines = <String>[];
  for (final c in root.children) {
    walk(c, 0, lines);
  }
  print(lines.join('\n'));
}

File: dump_tmpl_ast.mjs
// Dump official template AST in a canonical text format for comparison.
import { parse } from "@vue/compiler-dom";
import { readFile } from "node:fs/promises";

const samples = JSON.parse(await readFile("samples_tmpl.json", "utf8"));
const name = process.argv[2];
const s = samples.find((x) => x.name === name);
const m = s.sfc.match(/<template>([\s\S]*)<\/template>/);
const source = m[1];
const root = parse(source);
const enc = (t) => JSON.stringify(t);
const lines = [];
function loc(n) { return `@${n.loc.start.offset}-${n.loc.end.offset}`; }
function prop(p) {
  if (p.type === 6) {
    return `ATTR ${p.name} ${p.value ? enc(p.value.content) : "noval"} ${loc(p)}`;
  }
  const arg = p.arg ? `${p.arg.content}${p.arg.isStatic ? ":s" : ":d"}` : "-";
  const exp = p.exp ? p.exp.content : "-";
  const mods = p.modifiers.map((m) => m.content).join(",");
  return `DIR ${p.name} raw=${p.rawName} arg=${arg} exp=${enc(exp)} mods=[${mods}] ${loc(p)}`;
}
function walk(n, d) {
  const pad = "  ".repeat(d);
  if (n.type === 1) {
    lines.push(`${pad}EL ${n.tag} ns=${n.ns} tagType=${n.tagType} self=${n.isSelfClosing ? 1 : 0} ${loc(n)}`);
    for (const p of n.props) lines.push(`${pad}  ${prop(p)}`);
    for (const c of n.children) walk(c, d + 1);
  } else if (n.type === 2) {
    lines.push(`${pad}TEXT ${enc(n.content)} ${loc(n)}`);
  } else if (n.type === 3) {
    lines.push(`${pad}COMMENT ${enc(n.content)} ${loc(n)}`);
  } else if (n.type === 5) {
    lines.push(`${pad}INTERP ${enc(n.content.content)} ${loc(n)} exp@${n.content.loc.start.offset}-${n.content.loc.end.offset}`);
  }
}
for (const c of root.children) walk(c, 0);
console.log(lines.join("\n"));
